### ISSUE TITLE
Generating secondary solutions in same premake5.lua scripts is problematic

### ISSUE THREAD

**Issue by dgregorycryptic** [premake/premake-core#122]
I am attempting to make 2 solution files in my project.  One that contains all projects, and a second that contains a subset of the same projects.

Your wiki page on the subject indicates this is possible, but I'm not seeing a way through with the API as it is.

a) I can't specify the same project name without re-specifying all the details I already did further up in the same script.
b) The externalproject command requires a bunch of details. For instance, the uuid. And, the location.  Both of which are internally known already (or only).

There ought to be a way to refer to the existing project declaration. Do you have plans for a "projectref" command?

The upshot of what i'm asking for is to generate a second solution with references to already generated project files.

**Comment by dgregorycryptic** [comment]
Oh, I think I just stumbled on what you intended to be the solution for this.

The "includeexternal" external command.

**Comment by dgregorycryptic** [comment]
Hmm, it's not working completely.  Many projects I specify that are defined above in the same premake5.lua script (or included files) are not showing up in the .sln file  I will try to put together a small example the reproduces this problem.

**Comment by dgregorycryptic** [comment]
**premake5.lua**
```
solution "Main"
	configurations {"Debug", "Release"}
	platforms {"Win32", "x64"}
	include "A"
	include "B"
	include "C"
	include "D"

solution "Subset"
	configurations {"Debug", "Release"}
	platforms {"Win32", "x64"}
	includeexternal "A"
	includeexternal "C"
```
**A\premake5.lua**
```
project "A"
        kind "StaticLib"
        language "C++"
        buildinputs {"a.c"}
```
**B\premake5.lua**
```
project "B"
        kind "StaticLib"
        language "C++"
        buildinputs {"b.c"}
```
**C\premake5.lua**
```
project "C"
        kind "StaticLib"
        language "C++"
        buildinputs {"c.c"}
```
**D\premake5.lua**
```
project "D"
        kind "StaticLib"
        language "C++"
        buildinputs {"d.c"}
```
Output from premake5 (alpha4) run:
```
>premake5 vs2012
Building configurations...
Running action 'vs2012'...
Generating Main.sln...
Generating A.vcxproj...
Generating B.vcxproj...
Generating C.vcxproj...
Generating D.vcxproj...
Generating Subset.sln...
Done.
```
Contents of Main.sln, A-D.vcxproj are all correct.
Contents of **Subset.sln** are incorrect (no projects listed).
```
Microsoft Visual Studio Solution File, Format Version 12.00
# Visual Studio 2012
Global
	GlobalSection(SolutionConfigurationPlatforms) = preSolution
		Debug|Win32 = Debug|Win32
		Debug|x64 = Debug|x64
		Release|Win32 = Release|Win32
		Release|x64 = Release|x64
	EndGlobalSection
	GlobalSection(ProjectConfigurationPlatforms) = postSolution
	EndGlobalSection
	GlobalSection(SolutionProperties) = preSolution
		HideSolutionNode = FALSE
	EndGlobalSection
EndGlobal
```

Am I doing something wrong or is there a bug?

**Comment by TurkeyMan** [comment]
Who don't you just `include "a"` again in the second solution?
That's what I do...

**Comment by dgregorycryptic** [comment]
If I change the base **premake5.lua** to the following, as per your suggestion:

```
solution "Main"
    configurations {"Debug", "Release"}
    platforms {"Win32", "x64"}
    include "A"
    include "B"
    include "C"
    include "D"

solution "Subset"
    configurations {"Debug", "Release"}
    platforms {"Win32", "x64"}
    include "A"
    include "C"
```

...I get the exact same resulting Subset.sln. Still does not work.

**Comment by Meoo** [comment]
includeexternal calls include, and include is only called once per file, which means your two includes in the second solution are never executed. You could try dofile "A/premake5.lua" instead.

**Comment by dgregorycryptic** [comment]
Thank you, that does work.

But, I still believe this isn't working as intended, right?

**Comment by dgregorycryptic** [comment]
There is one downside to regenerating the "A" or "C" projects again (via 'dofile "A/premake.lua" ').  If the second solution contains less configurations or platforms that the first inclusion, it will overwrite that file and remove what you generated before.  This same problem will exist for any other options that are changed in your script between the first and second solutions.

**Comment by starkos** [comment]
This is where you need externals. What about:

```lua
solution "Main"
    configurations {"Debug", "Release"}
    platforms {"Win32", "x64"}
    include "A"
    include "B"
    include "C"
    include "D"

solution "Subset"
    configurations {"Debug", "Release"}
    platforms {"Win32", "x64"}
    externalproject "A"
    externalproject "C"
```

**Comment by dgregorycryptic** [comment]
externalproject "A" on its own won't work.  You need to **repeat** at least some of the information in A/premake5.lua, like "kind".

```
Building configurations...
Error: ** Error: project 'A' needs a kind in configuration 'Debug|Win32'
```

There are other errors behind that.

This is what I believe you designed includeexternal for - it just doesn't work.

**Comment by dgregorycryptic** [comment]
Looking at globals.lua, the implementation of function include(), which includexternal() calls, does tracking of loaded premake5.lua files and only allows each to load once.  This is why calling includeexternal (or include directly) does absolutely nothing the second time.

I really don't know how to do what I want right now.  Totally blocked.

**Comment by dgregorycryptic** [comment]
Unblocked with the following override:

```
function forceincludeexternal(fname)
	local fullPath = os.locate(fname, fname .. ".lua", path.join(fname, "premake5.lua"), path.join(fname, "premake4.lua"))
	premake.api._isIncludingExternal = true
	fname = fullPath or fname
	dofile(fname)
	premake.api._isIncludingExternal = nil
end

```

**Comment by starkos** [comment]
Yeah, this is a good point. I can see the value of adding a `doexternal()` call that does the same work without the was-previously-loaded check, definitely. I'm wondering if we should just get rid of `includeexternal()` entirely to avoid this confusion in the future, and if it would be better if I used `dofile()` instead of `include()` in the examples.

**Comment by dgregorycryptic** [comment]
If you want the semantics of includeexternal to match that of include (a single execution of premake actually processes the named file once and only once), then I'm not honestly sure of the usefulness of the command.  It will only operate the way you'd expect the first time it's encountered for a particular path.

forceincludeexternal (the way I wrote it) doesn't generate anything downstream, but does let references to it be generated.  One could argue that's really the functionality that is needed.

I would hate to lose the api._isIncludingExternal honoring inside the engine, because then I couldn't accomplish what I'm trying to accomplish.  Not sure if that's what you're meaning by "get rid of internexternal() entirely".

**Comment by starkos** [comment]
Talking about replacing it with a new call `doexternal()` which will always run the file (skipping the "is already loaded" check) while still marking new containers as external. Calling it `doexternal()` to align it with `dofile()` rather than `include()` (which is still useful on its own, I think).

**Comment by dgregorycryptic** [comment]
I see.  I do like the semantics where it will concat premake5.lua on the include path.  So I sorta wish that could stay.  But it's definitely up to you. This is only feedback.

**Comment by starkos** [comment]
> I do like the semantics where it will concat premake5.lua on the include path

Yes, that's important. Should we add it to `dofile()`? I'd really rather not have yet another way to load scripts if it can be helped. Or is removing the previously-loaded check from `includeexternal()` the way to go? I don't know how much people rely on the loading check.

**Comment by dgregorycryptic** [comment]
I view dofile() as a fairly low level interface.  I wouldn't muck with that, personally.

I would prefer to see includeexternal() do what's its supposed to (of course, my opinion) - remove the previously-loaded checks.

Look at it this way - what happens if you remove those previously-loaded checks?  I would assume what we'd want is that no additional **files** (.sln or .vcxproj) would be generated, but **references** to those files would be.

I can't speak to other generators - not as familiar with them.

Perhaps this is a better question for the google groups list.

**Comment by starkos** [comment]
I agree. Can you submit a pull request with that change, and I will update the documentation to note the difference between the calls once it lands?

**Comment by dgregorycryptic** [comment]
I have not actually pushed any code fix to github for this.  At this point, it's in my overrides file only.  I will try to spend some time formalizing it, but I'm on a tight deadline.  (aren't we all?)

**Comment by starkos** [comment]
Cool, thanks. You can consider this on my to-do list as well, but I'm not sure when I'll get to it…you know, same old story :)

**Comment by samsinsane** [comment]
@starkos I believe this can also be resolved, as #228 was merged. :)
